In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

import pandas
from sklearn.model_selection import train_test_split
import numpy as np
import time
# from sklearn.externals import joblib

from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

### Import the necessary libraries and read the provided CSVs as a data frame and perform the below steps.

1. Merge the provided CSVs into one data-frame

In [2]:
# Importing all the given datasets
data1 = pd.read_csv("phone_user_review_file_1.csv")
data2 = pd.read_csv("phone_user_review_file_2.csv")
data3 = pd.read_csv("phone_user_review_file_3.csv")
data4 = pd.read_csv("phone_user_review_file_4.csv")
data5 = pd.read_csv("phone_user_review_file_5.csv")
data6 = pd.read_csv("phone_user_review_file_6.csv")

In [3]:
# Since all columns and datatypes are same, appending the dataframes to one
df = data1.append([data2, data3, data4, data5, data6])

2. Check a few observations and shape of the data-frame

In [4]:
# Check for data info
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1415133 entries, 0 to 163836
Data columns (total 11 columns):
 #   Column     Non-Null Count    Dtype  
---  ------     --------------    -----  
 0   phone_url  1415133 non-null  object 
 1   date       1415133 non-null  object 
 2   lang       1415133 non-null  object 
 3   country    1415133 non-null  object 
 4   source     1415133 non-null  object 
 5   domain     1415133 non-null  object 
 6   score      1351644 non-null  float64
 7   score_max  1351644 non-null  float64
 8   extract    1395772 non-null  object 
 9   author     1351931 non-null  object 
 10  product    1415132 non-null  object 
dtypes: float64(2), object(9)
memory usage: 129.6+ MB


In [5]:
df.head(5)

phone_url       date lang country            source  \
0  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us  Verizon Wireless   
1  /cellphones/samsung-galaxy-s8/  4/28/2017   en      us       Phone Arena   
2  /cellphones/samsung-galaxy-s8/   5/4/2017   en      us            Amazon   
3  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us           Samsung   
4  /cellphones/samsung-galaxy-s8/  5/11/2017   en      us  Verizon Wireless   

                domain  score  score_max  \
0  verizonwireless.com   10.0       10.0   
1       phonearena.com   10.0       10.0   
2           amazon.com    6.0       10.0   
3          samsung.com    9.2       10.0   
4  verizonwireless.com    4.0       10.0   

                                             extract       author  \
0  As a diehard Samsung fan who has had every Sam...   CarolAnn35   
1  Love the phone. the phone is sleek and smooth ...    james0923   
2  Adequate feel. Nice heft. Processor's still sl...     R. Craig   
3  Never disappointed. One of the reasons I've be...  Buster2020    
4  I've now found that i'm in a group of people t...   S Ate Mine   

                                             product  
0                                  Samsung Galaxy S8  
1                                  Samsung Galaxy S8  
2  Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                      Samsung Galaxy S8 64GB (AT&T)  
4                                  Samsung Galaxy S8

3. Round off scores to the nearest integers

In [6]:
df.score = df['score'].round()

4. Check for missing values. Impute missing values if there are any

In [7]:
df.isnull().sum().sum() 

209542

In [8]:
# Identifying the percentage of missing values for each columns
percent_missing = df.isnull().sum() #* 100 / len(df)
missing_value_df = pd.DataFrame({'column_name': df.columns,
                                 'percent_missing': percent_missing})

#Sorting the dataframe based on percentage of missing values
missing_value_df.sort_values('percent_missing', inplace=True)
missing_value_df

column_name  percent_missing
phone_url   phone_url                0
date             date                0
lang             lang                0
country       country                0
source         source                0
domain         domain                0
product       product                1
extract       extract            19361
author         author            63202
score           score            63489
score_max   score_max            63489

In [9]:
data = df.dropna()

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1275917 entries, 0 to 163836
Data columns (total 11 columns):
 #   Column     Non-Null Count    Dtype  
---  ------     --------------    -----  
 0   phone_url  1275917 non-null  object 
 1   date       1275917 non-null  object 
 2   lang       1275917 non-null  object 
 3   country    1275917 non-null  object 
 4   source     1275917 non-null  object 
 5   domain     1275917 non-null  object 
 6   score      1275917 non-null  float64
 7   score_max  1275917 non-null  float64
 8   extract    1275917 non-null  object 
 9   author     1275917 non-null  object 
 10  product    1275917 non-null  object 
dtypes: float64(2), object(9)
memory usage: 116.8+ MB


5. Check for duplicate values and remove them if there is any.

In [11]:
data.drop_duplicates()

phone_url       date lang country  \
0          /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
1          /cellphones/samsung-galaxy-s8/  4/28/2017   en      us   
2          /cellphones/samsung-galaxy-s8/   5/4/2017   en      us   
3          /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
4          /cellphones/samsung-galaxy-s8/  5/11/2017   en      us   
...                                   ...        ...  ...     ...   
163832  /cellphones/alcatel-ot-club_1187/  5/12/2000   de      de   
163833  /cellphones/alcatel-ot-club_1187/  5/11/2000   de      de   
163834  /cellphones/alcatel-ot-club_1187/   5/4/2000   de      de   
163835  /cellphones/alcatel-ot-club_1187/   5/1/2000   de      de   
163836  /cellphones/alcatel-ot-club_1187/  4/25/2000   de      de   

                  source               domain  score  score_max  \
0       Verizon Wireless  verizonwireless.com   10.0       10.0   
1            Phone Arena       phonearena.com   10.0       10.0   
2                 Amazon           amazon.com    6.0       10.0   
3                Samsung          samsung.com    9.0       10.0   
4       Verizon Wireless  verizonwireless.com    4.0       10.0   
...                  ...                  ...    ...        ...   
163832              Ciao              ciao.de    2.0       10.0   
163833              Ciao              ciao.de   10.0       10.0   
163834              Ciao              ciao.de    2.0       10.0   
163835              Ciao              ciao.de    8.0       10.0   
163836              Ciao              ciao.de    2.0       10.0   

                                                  extract        author  \
0       As a diehard Samsung fan who has had every Sam...    CarolAnn35   
1       Love the phone. the phone is sleek and smooth ...     james0923   
2       Adequate feel. Nice heft. Processor's still sl...      R. Craig   
3       Never disappointed. One of the reasons I've be...   Buster2020    
4       I've now found that i'm in a group of people t...    S Ate Mine   
...                                                   ...           ...   
163832  Weil mein Onkel bei ALcatel arbeitet habe ich ...    david.paul   
163833  Hy Liebe Leserinnen und Leser!! Ich habe seit ...  Christiane14   
163834  Jetzt hat wohl Alcatell gedacht ,sie machen wa...    michaelawr   
163835  Ich bin seit 2 Jahren (stolzer) Besitzer eines...   claudia0815   
163836  Was sich Alkatel hier wieder ausgedacht hat,sc...    michaelawr   

                                                  product  
0                                       Samsung Galaxy S8  
1                                       Samsung Galaxy S8  
2       Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                           Samsung Galaxy S8 64GB (AT&T)  
4                                       Samsung Galaxy S8  
...                                                   ...  
163832                            Alcatel Club Plus Handy  
163833                            Alcatel Club Plus Handy  
163834                            Alcatel Club Plus Handy  
163835                            Alcatel Club Plus Handy  
163836                            Alcatel Club Plus Handy  

[1271437 rows x 11 columns]

6. Keep only 1000000 data samples. Use random state=612.

In [12]:
data = data.sample(n=1000000, replace=True, random_state=612)
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000000 entries, 146164 to 58871
Data columns (total 11 columns):
 #   Column     Non-Null Count    Dtype  
---  ------     --------------    -----  
 0   phone_url  1000000 non-null  object 
 1   date       1000000 non-null  object 
 2   lang       1000000 non-null  object 
 3   country    1000000 non-null  object 
 4   source     1000000 non-null  object 
 5   domain     1000000 non-null  object 
 6   score      1000000 non-null  float64
 7   score_max  1000000 non-null  float64
 8   extract    1000000 non-null  object 
 9   author     1000000 non-null  object 
 10  product    1000000 non-null  object 
dtypes: float64(2), object(9)
memory usage: 91.6+ MB


7. Drop irrelevant features. Keep features like Author, Product, and Score.

In [13]:
remove_col = ['phone_url','date','lang','country','domain','extract','score_max','source']

In [14]:
# Remove those columns/features from the main dataset
data = data.drop(remove_col, axis = 1)
data.shape
data.head(4)

(1000000, 3)

score                              author  \
146164   10.0                          titanicb52   
194012    2.0                     Amazon Customer   
256524    6.0                               Nancy   
37940     8.0  Jonas "Description: Indescribable"   

                                                  product  
146164                                               7135  
194012                            BlackBerry PRIV (Black)  
256524  Alcatel OneTouch Idol 3 Global Unlocked 4G LTE...  
37940   Samsung Galaxy Nexus I515 Camera Touch Android...

### Answer the following questions

1. Identify the most rated features.

In [15]:
data['product'].value_counts()

Lenovo Vibe K4 Note (White,16GB)              4104
Lenovo Vibe K4 Note (Black, 16GB)             3483
OnePlus 3 (Graphite, 64 GB)                   3177
OnePlus 3 (Soft Gold, 64 GB)                  2852
Huawei P8lite zwart / 16 GB                   2088
                                              ... 
HTC S711 Painel frontal com lente                1
HUAWEI P9 32GB DS PLATA                          1
LG C2000                                         1
Sony Ericsson K550iM                             1
Steelseries Siberia P300 ‐kuulokemikrofoni       1
Name: product, Length: 47617, dtype: int64

#### Comment : The product 'Lenovo Vibe K4 Note (White,16GB) is the one that has been rated the most number of times

2. Identify the users with most number of reviews.

In [16]:
data['author'].value_counts()

Amazon Customer    60098
Cliente Amazon     15243
e-bit               6646
Client d'Amazon     6162
Amazon Kunde        3804
                   ...  
emily souza            1
Amber Butler           1
rufus                  1
Richu                  1
lisakristiina          1
Name: author, Length: 453418, dtype: int64

#### Comment : A unknown user marked as 'Amazon Customer' is the one who has reviewed the most number of times

3. Select the data with products having more than 50 ratings and users who have given more than 50 ratings. Report the shape of the final dataset.

In [17]:
data.describe()

score
count  1000000.000000
mean         8.011225
std          2.622101
min          0.000000
25%          8.000000
50%          9.000000
75%         10.000000
max         10.000000

In [18]:
counts1 = pd.value_counts(data['product'])
counts2 = pd.value_counts(data['author'])

In [19]:
data1 = data[(data['product'].isin(counts1[counts1 >= 50].index))&(data['author'].isin(counts2[counts2 >= 50].index))]

In [20]:
data1.shape

(113509, 3)

### Build a popularity based model and recommend top 5 mobile phones.

1. Count of author for each unique product as recommendation score 

In [21]:
data1.head(3)

score           author  \
194012    2.0  Amazon Customer   
256524    6.0            Nancy   
332575    8.0           Pietro   

                                                  product  
194012                            BlackBerry PRIV (Black)  
256524  Alcatel OneTouch Idol 3 Global Unlocked 4G LTE...  
332575  Lenovo Motorola Moto G Smartphone, 4,5 pollici...

In [22]:
train_data, test_data = train_test_split(data1, test_size = 0.25, random_state=123)

In [23]:
#Count of author for each unique product as recommendation score 
train_data_grouped = train_data.groupby('product').agg({'author': 'count'}).reset_index()
train_data_grouped.rename(columns = {'author': 'score'},inplace=True)
train_data_grouped.head()

product  score
0                                               3220      2
1  5.5-Inch Unlocked Lenovo A850 3G Smartphone-(9...      5
2  AICEK Coque ASUS ZenFone 3 Max ZC520TL, AICEK ...     18
3  AICEK Coque ASUS ZenFone 3 ZE520KL, AICEK Etui...     15
4  AICEK Coque Samsung Galaxy A3 2016, AICEK Etui...     63

2. Sort the products on recommendation score 

In [24]:
#Sort the songs on recommendation score 
train_data_sort = train_data_grouped.sort_values(['score', 'product'], ascending = [0,1]) 
      
#Generate a recommendation rank based upon score 
train_data_sort['Rank'] = train_data_sort['score'].rank(ascending=0, method='first') 

3. Get the top 5 recommendations

In [25]:
#Get the top 5 recommendations 
popularity_recommendations = train_data_sort.head(5) 
popularity_recommendations 

product  score  Rank
1607     Lenovo Vibe K4 Note (White,16GB)   1774   1.0
1606    Lenovo Vibe K4 Note (Black, 16GB)   1513   2.0
2457          OnePlus 3 (Graphite, 64 GB)   1059   3.0
2458         OnePlus 3 (Soft Gold, 64 GB)   1052   4.0
1608  Lenovo Vibe K5 (Gold, VoLTE update)    870   5.0

### Build a collaborative filtering model using SVD. You can use SVD from surprise or build it from scratch(Note: Incase you’re building it from scratch.
[ you can limit your data points to 5000 samples if you face memory issues). Build a collaborative filtering model using kNNWithMeans from surprise. You can try both user-based and item-based model.]

In [26]:
# !pip install scikit-surprise
from surprise import Dataset,Reader
from surprise.model_selection import cross_validate
from surprise import NormalPredictor
from surprise.model_selection import train_test_split  
from surprise import SVD, KNNWithMeans
from surprise import accuracy
from surprise import KNNWithMeans
from surprise import accuracy
from surprise.model_selection import RandomizedSearchCV

In [27]:
reader = Reader(rating_scale=(1, 10))

data_new = Dataset.load_from_df(data1[['author', 'product', 'score']], reader)
trainset, testset = train_test_split(data_new, test_size=.3, random_state=10)

In [28]:
# Split data to train and test
trainset, testset = train_test_split(data_new, test_size=.25,random_state=123)

In [29]:
trainset.all_ratings()

<generator object Trainset.all_ratings at 0x0000028093C49040>

1. Collaborative filter using SVD

In [30]:
svd_model = SVD(n_factors=5,biased=False)
svd_model.fit(trainset)

In [31]:
testset[0]

('Amazon Customer', 'HTC One Me dual sim (32GB, Meteor Grey)', 10.0)

In [32]:
test_pred = svd_model.test(testset)

In [33]:
accuracy.rmse(test_pred)

RMSE: 2.6723


2.6722538704120105

### KNN and SVD Best Model fit using Cross Validation

In [34]:
from collections import defaultdict

def get_top_n(predictions, n=10):
    # First map the predictions to each user.
    top_n = defaultdict(list)
    for uid, iid, true_r, est, _ in predictions:
        top_n[uid].append((iid, est))

    # Then sort the predictions for each user and retrieve the k highest ones.
    for uid, user_ratings in top_n.items():
        user_ratings.sort(key=lambda x: x[1], reverse=True)
        top_n[uid] = user_ratings[:n]

    return top_n

In [35]:
class collab_filtering_based_recommender_model():
    def __init__(self, model, trainset, testset, data):
        self.model = model
        self.trainset = trainset
        self.testset = testset
        self.data = data
        self.pred_test = None
        self.recommendations = None
        self.top_n = None
        self.recommenddf = None
        
    def fit_and_predict(self):      
        self.model.fit(self.trainset)       
        self.pred_test = self.model.test(self.testset)        
        rmse = round(accuracy.rmse(self.pred_test), 3)
        print('RMSE for the predicted result is ' + str(rmse))   
        
        self.top_n = get_top_n(self.pred_test)
        self.recommenddf = pd.DataFrame(columns=['userId', 'productId', 'Rating'])
        for item in self.top_n:
            subdf = pd.DataFrame(self.top_n[item], columns=['productId', 'Rating'])
            subdf['userId'] = item
            cols = subdf.columns.tolist()
            cols = cols[-1:] + cols[:-1]
            subdf = subdf[cols]        
            self.recommenddf = pd.concat([self.recommenddf, subdf], axis = 0)        
        return rmse
        
    def cross_validate(self):
        cv_result = cross_validate(self.model, self.data, n_jobs=-1)
        cv_result = round(cv_result['test_rmse'].mean(),3)
        print('Mean CV RMSE is ' + str(cv_result))
        return cv_result
    
    def recommend(self, user_id, n=5):
        print('Recommending top ' + str(n)+ ' products for userid : ' + user_id + ' ...**')
        df = self.recommenddf[self.recommenddf['userId'] == user_id].head(n)
        display(df)
        return df

In [36]:
def find_best_model(model, parameters,data):
    clf = RandomizedSearchCV(model, parameters, n_jobs=-1, measures=['rmse'])
    clf.fit(data)             
    print(clf.best_score)
    print(clf.best_params)
    print(clf.best_estimator)
    return clf

#### User Based : Cross Validate to find the best KNN model parameters

In [38]:
sim_options = {
    "name": ["msd", "cosine", "pearson"],
    "user_based": [True]
}
params = { 'k': range(10,20,2), 'sim_options': sim_options}
clf = find_best_model(KNNWithMeans, params, data_new)

{'rmse': 2.57489734020333}
{'rmse': {'k': 18, 'sim_options': {'name': 'msd', 'user_based': True}}}
{'rmse': <surprise.prediction_algorithms.knns.KNNWithMeans object at 0x00000280FE9819A0>}


In [39]:
knnwithmeans = clf.best_estimator['rmse']
col_fil_knn_user = collab_filtering_based_recommender_model(knnwithmeans, trainset, testset, data_new)

In [40]:
knnwithmeans_rmse = col_fil_knn_user.fit_and_predict()

Computing the msd similarity matrix...
Done computing similarity matrix.
RMSE: 2.5942
RMSE for the predicted result is 2.594


In [41]:
knnwithmeans_cv_rmse = col_fil_knn_user.cross_validate()

Mean CV RMSE is 2.566


In [42]:
result_knn_user1 = col_fil_knn_user.recommend(user_id='Justin', n=5)
result_knn_user1 = col_fil_knn_user.recommend(user_id='Sebastian', n=5)
result_knn_user1 = col_fil_knn_user.recommend(user_id='Sarah', n=5)

Recommending top 5 products for userid : Justin ...**


userId                                          productId    Rating
0  Justin               Samsung S5830i Galaxy Ace Onyx Black      10.0
1  Justin               Samsung S5830i Galaxy Ace Onyx Black      10.0
2  Justin               Samsung S5830i Galaxy Ace Onyx Black      10.0
3  Justin  Apple iPhone 6s Plus Unlocked GSM 4G LTE Smart...  9.639526
4  Justin  Samsung Galaxy Note II N7100 16GB Gray-Unlocke...  9.327505

Recommending top 5 products for userid : Sebastian ...**


userId                                          productId    Rating
0  Sebastian  Lenovo Motorola RAZR Smartphone (10,9 cm (4,3 ...  9.971946
1  Sebastian  Wiko Rainbow Smartphone (12,7 cm (5 Zoll) Disp...  9.584584
2  Sebastian  HTC One Smartphone (11,9 cm (4,7 Zoll) Touchsc...  9.578469
3  Sebastian  LG Nexus 5 Smartphone, memoria 16GB, Android 4...  9.539701
4  Sebastian  Huawei Ascend Y530 Smartphone (4,5 Zoll (11,4 ...  9.444705

Recommending top 5 products for userid : Sarah ...**


userId                                          productId    Rating
0  Sarah  Motorola Moto E 2nd Generation 4G LTE White XT...      10.0
1  Sarah                         Microsoft Nokia 3310 Handy  9.192737
2  Sarah  HTC Desire 816 Black (Virgin mobile) - 5.5 inc...   9.19011
3  Sarah  Samsung Galaxy J5 2016 Sim Free Mobile Phone -...   8.97479
4  Sarah            Nokia 3310 - SIM Free (various colours)  8.974328

#### Item Based : Cross Validate to find the best KNN model parameters

In [55]:
# Limiting data size because of the extreme loading issues in the cross validation step
data2 = data1.sample(n=10000, replace=True, random_state=612)

reader = Reader(rating_scale=(1, 10))

data_new1 = Dataset.load_from_df(data2[['author', 'product', 'score']], reader)
trainset, testset = train_test_split(data_new1, test_size=.3, random_state=10)

In [56]:
sim_options = {
    "name": ["msd", "cosine","pearson"],
    "user_based": [False]
}
params = { 'k': range(10,20,1), 'sim_options': sim_options}
clf = find_best_model(KNNWithMeans, params, data_new1)

{'rmse': 2.9254182806295623}
{'rmse': {'k': 15, 'sim_options': {'name': 'cosine', 'user_based': False}}}
{'rmse': <surprise.prediction_algorithms.knns.KNNWithMeans object at 0x00000280C50BD280>}


In [57]:
knnwithmeans = clf.best_estimator['rmse']
col_fil_knn_item = collab_filtering_based_recommender_model(knnwithmeans, trainset, testset, data_new1)

In [58]:
knnwithmeans_rmse = col_fil_knn_item.fit_and_predict()

Computing the cosine similarity matrix...
Done computing similarity matrix.
RMSE: 2.9364
RMSE for the predicted result is 2.936


In [59]:
knnwithmeans_cv_rmse = col_fil_knn_item.cross_validate()

Mean CV RMSE is 2.935


In [62]:
result_knn_user1 = col_fil_knn_item.recommend(user_id='Henry', n=5)
result_knn_user1 = col_fil_knn_item.recommend(user_id='Dani', n=5)
result_knn_user1 = col_fil_knn_item.recommend(user_id='Sarah', n=5)

Recommending top 5 products for userid : Henry ...**


userId                                          productId     Rating
0  Henry  Asus ZenFone 3 Laser 5.5-inch Glacier Silver [...  10.000000
1  Henry  LG G4 H815 - Smartphone de 5.5" (4G, Qualcomm ...  10.000000
2  Henry  HTC One M7 Unlocked GSM 4G LTE Quad-Core Smart...   7.783286
3  Henry    Samsung GT-E1270 Téléphone portable 32 Mo Rouge   7.783286

Recommending top 5 products for userid : Dani ...**


userId                                          productId     Rating
0   Dani  LG G4 H815 - Smartphone de 5.5" (4G, Qualcomm ...  10.000000
1   Dani        LG D802 G2 Smartphone, 16 GB, Nero [Italia]   8.000000
2   Dani  Lenovo Moto G4 Play - Smartphone de 5 (4G, RAM...   6.547619
3   Dani  Lenovo Moto G4 Play - Smartphone de 5 (4G, RAM...   6.547619

Recommending top 5 products for userid : Sarah ...**


userId                                          productId    Rating
0  Sarah                        Sony Xperia X zwart / 32 GB  9.000000
1  Sarah  LG Electronics GM360 Viewty Plus Smartphone (7...  7.783286
2  Sarah      BLU Studio 5.0S II Quad Band Unlocked (White)  7.783286
3  Sarah                   Apple iPhone 7, 32Gb - Rose Gold  7.783286
4  Sarah        Sim Free iPhone 7 32GB Mobile Phone - Black  7.783286

#### Cross Validate SVD to find the best model parameters

In [63]:
params= {
    "n_epochs": [5, 10, 15, 20],
    "lr_all": [0.002, 0.005],
    "reg_all": [0.4, 0.6]
}
clf = find_best_model(SVD, params, data_new)

{'rmse': 2.591862504164225}
{'rmse': {'n_epochs': 15, 'lr_all': 0.005, 'reg_all': 0.4}}
{'rmse': <surprise.prediction_algorithms.matrix_factorization.SVD object at 0x00000280C5143190>}


In [64]:
svd = clf.best_estimator['rmse']
col_fil_svd = collab_filtering_based_recommender_model(svd, trainset, testset, data_new)

In [65]:
svd_rmse = col_fil_svd.fit_and_predict()

RMSE: 2.6675
RMSE for the predicted result is 2.667


In [66]:
svd_cv_rmse = col_fil_svd.cross_validate()

Mean CV RMSE is 2.595


In [67]:
result_svd_user1 = col_fil_svd.recommend(user_id='Henry', n=5)
result_svd_user1 = col_fil_svd.recommend(user_id='Dani', n=5)
result_svd_user1 = col_fil_svd.recommend(user_id='Sarah', n=5)

Recommending top 5 products for userid : Henry ...**


userId                                          productId    Rating
0  Henry  Asus ZenFone 3 Laser 5.5-inch Glacier Silver [...  7.751462
1  Henry  LG G4 H815 - Smartphone de 5.5" (4G, Qualcomm ...  7.723790
2  Henry  HTC One M7 Unlocked GSM 4G LTE Quad-Core Smart...  7.572664
3  Henry    Samsung GT-E1270 Téléphone portable 32 Mo Rouge  7.572664

Recommending top 5 products for userid : Dani ...**


userId                                          productId    Rating
0   Dani  Lenovo Moto G4 Play - Smartphone de 5 (4G, RAM...  8.676137
1   Dani  Lenovo Moto G4 Play - Smartphone de 5 (4G, RAM...  8.676137
2   Dani  LG G4 H815 - Smartphone de 5.5" (4G, Qualcomm ...  8.096557
3   Dani        LG D802 G2 Smartphone, 16 GB, Nero [Italia]  7.937554

Recommending top 5 products for userid : Sarah ...**


userId                                          productId    Rating
0  Sarah                        Sony Xperia X zwart / 32 GB  6.763089
1  Sarah  LG Electronics GM360 Viewty Plus Smartphone (7...  6.732305
2  Sarah      BLU Studio 5.0S II Quad Band Unlocked (White)  6.732305
3  Sarah                   Apple iPhone 7, 32Gb - Rose Gold  6.732305
4  Sarah        Sim Free iPhone 7 32GB Mobile Phone - Black  6.732305

### Report your findings and inferences

* For Sarah, item based and SVD gave almost the same recommendations
* Item - item collaborative filtering offers the worst RMSE ( constraint : data size is reduced because of computation issues)
* Each user has different product being recommended unlike popularity based filtering
* An RMSE score of 2.59 is the best available from the model - User User Collaborative filtering

### In what business scenario you should use popularity based Recommendation Systems ?

* Apps like Spotify, Amazon music etc to recommend the trending songs of the seasons
* Websites of small businesses that does not have a user profile hence recommends the fastest moving products
* YouTube trending videos or Google popular/trending news section
* Swiggy/Zomato suggesting the most popular restaurants and dishes

### In what business scenario you should use CF based Recommendation Systems ?

* Amazon's recommender system - Also Myntra/Flipkart where items are suggested based on user preference or items often bought together
* Facebook suggesting common friends
* Netflix suggesting movies/shows based on the last watched TV shows

### What other possible methods can you think of which can further improve the recommendation for different users ?

* Cold start problems to be addressed with a fall back scenario that recommends popular products for first time users
* System detects if it has enough data to personalize recommendations to a certain customer
* This leads to hybrid recommendation systems
* Content based collaborative filtering is another method for recommening based on item specifications